content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+You might notice that the interface is not very responsive. To amplify the issue, I artificially slowed down the rendering time of the names list. And since this list is big, it has a significant impact on the application’s performance. This is not good 😰.
+
+Our users expect immediate feedback, but the app is unresponsive for seconds after a keystroke. To understand what’s going on, let’s take a look at the DevTools’ Performance tab. Here’s a screenshot of a recording while I type the word "Ada" into the search box:
+
+![Screenshot of Chrome DevTools that shows that the three keypress events take 733ms to render.](devtools-sync.png)
+
+We can see there are a lot of red triangles, which is usually not a good sign. For every keystroke, we see a `keypress` event being fired. All three events run within one frame,[^4] causing that frame to take **733ms**. That’s way above our average frame budget of 16ms.
+
+Inside this `keypress` event, our React code will be called, which causes the input value and the search value to update and then send the analytics notification. In turn, the updated state values will cause the app to re-render–down to every individual name. That’s quite a lot of work that we have to do that will block the main thread!
+
+The first step to improve the status quo is to enable the unstable Concurrent Mode. This can be done by wrapping a part of our React tree with the `<React.unstable_ConcurrentMode>` component, like this[^3]:
+
+```diff
+- ReactDOM.render(<App />, container);
++ ReactDOM.render(
++  <React.unstable_ConcurrentMode>
++    <App />
++  </React.unstable_ConcurrentMode>,
++  rootElement
++ );
+```
+
+However, enabling Concurrent Mode alone will not change the experience in our case. React will still receive both state updates at the same time, so there’s no way of knowing which is less important.
+
+We instead want to set the input value first so that we only need to update the search box in the beginning. Updates to the search value and our analytics notification should happen later. To do this, we’re using an API exposed by the Scheduler package (which can be installed with `npm i scheduler`) to enqueue a lower priority callback:
+
+```js{1,10-14}
+import { unstable_next } from "scheduler";
+
+function SearchBox(props) {
+  const [inputValue, setInputValue] = React.useState();
+
+  function handleChange(event) {
+    const value = event.target.value;
+
+    setInputValue(value);
+    unstable_next(function() {
+      props.onChange(value);
+      sendAnalyticsNotification(value);
+    });
+  }
+
+  return <input type="text" value={inputValue} onChange={handleChange} />;
+}
+```
+
+The API we’re using, `unstable_next()`, will run the callback with the `Normal` priority. Indeed, our input box already feels a lot more responsive, and frames no longer get dropped while typing into it.
```

Comment:
I'm guessing that what the scheduler API does is check a callback list while in a `requestAnimationFrame` callback, run any of the callbacks that will timeout otherwise and, if any time is left from the 16ms budget, run the rest in priority order, is that so? (In any case, it would be great if you could elaborate a bit on how it works internally)

Reply:
See https://github.com/philipp-spiess/philipp-spiess.github.io/pull/1#discussion_r262226909

The actual callbacks are not checked in rAF since rAF means "frame needs to render soon" instead, it will schedule a postMessage callback that fires "right after" rAF is completed and the frame is rendered and then check the registered callbacks and their timeout in the postMessage callback. 

I'm not sure if this is something that we want to explain especially since it's subject to change and requires quite some understanding of the underlying DOM APIs.